File: app/src/main/java/com/mycompany/myapp/ui/main/MainComponent.java
package com.mycompany.myapp.ui.main;

import android.content.Context;

import com.mycompany.myapp.data.api.github.GitHubService;
import com.mycompany.myapp.ui.ActivityScope;
import com.mycompany.myapp.ui.main.MainComponent.MainModule;

import javax.inject.Named;

import dagger.Module;
import dagger.Provides;
import dagger.Subcomponent;
import io.reactivex.Scheduler;

@ActivityScope
@Subcomponent(modules = MainModule.class)
public interface MainComponent {

    @Module
    class MainModule {
        private final MainActivity activity;

        MainModule(MainActivity activity) {
            this.activity = activity;
        }

        @ActivityScope
        @Provides
        protected MainPresenter providePresenter(
                Context context,
                GitHubService service,
                @Named("io") Scheduler ioScheduler,
                @Named("main") Scheduler mainScheduler,
                @Named("loading_delay_ms") long loadingDelayMs) {
            return new MainPresenter(context, service, ioScheduler, mainScheduler, loadingDelayMs);
        }
    }

    void inject(MainActivity activity);
    void inject(MainFragment fragment);
}

File: app/src/main/java/com/mycompany/myapp/ui/main/MainComponent.kt
package com.mycompany.myapp.ui.main

import android.content.Context

import com.mycompany.myapp.data.api.github.GitHubInteractor
import com.mycompany.myapp.ui.ActivityScope
import com.mycompany.myapp.ui.main.MainComponent.MainModule

import dagger.Module
import dagger.Provides
import dagger.Subcomponent
import io.reactivex.Scheduler
import javax.inject.Named

@ActivityScope
@Subcomponent(modules = arrayOf(MainModule::class))
interface MainComponent {

    @Module
    class MainModule internal constructor(private val activity: MainActivity) {

        @ActivityScope
        @Provides
        fun providePresenter(
                context: Context, interactor:
                GitHubInteractor,
                @Named("io") ioScheduler: Scheduler,
                @Named("main") mainScheduler: Scheduler,
                @Named("loading_delay_ms") loadingDelayMs: Long): MainPresenter {
            return MainPresenter(
                    context,
                    interactor,
                    ioScheduler,
                    mainScheduler,
                    loadingDelayMs)
        }
    }

    fun inject(activity: MainActivity)
    fun inject(fragment: MainFragment)
}
